src/diagnostics.jl

```diff
@@ -55,25 +52,174 @@ function velocity_div!(grid::RegularCartesianGrid, u, v, w, div)
-        println("WARNING: Velocity divergence is high! min=$min_div, mean=$mean_div, max=$max_div")
+        if col == cols
+            sum = 0
+            for j in 1:cols
+                sum += Rx[1, j, lvl]
+            end
+            Rxy[1, 1, lvl] = real(sum)
+        end
     end
 
-    if max(abs(min_div), abs(max_div)) >= diag.abort_threshold
-        t, i = model.clock.time, model.clock.iteration
-        println("time = $t, iteration = $i")
-        println("Velocity divergence is too high! min=$min_div, mean=$mean_div, max=$max_div. Aborting simulation.")
-        exit(1)
+    return
+end
+
+####
+#### Horizontally averaged vertical profiles
+####
+
+mutable struct VerticalProfile{P, F, I, T} <: Diagnostic
+      profile :: P
+        field :: F
+    frequency :: I
+     interval :: T
+     previous :: Float64
+end
+
+function VerticalProfile(model, field; frequency=nothing, interval=nothing)
+    freq, int = validate_interval(frequency, interval)
+    profile = zeros(model.arch, model.grid, 1, 1, model.grid.Nz)
+    VerticalProfile(profile, field, freq, int, 0.0)
+end
+
+function run_diagnostic(model::Model, P::VerticalProfile{<:Array})
+    P.profile .= mean(data(P.field), dims=[1, 2])
+end
+
+@hascuda function run_diagnostic(model::Model, P::VerticalProfile{<:CuArray})
+    Nx, Ny, Nz = P.field.grid.Nx, P.field.grid.Ny, P.field.grid.Nz
+    sz = 2Nx * sizeof(eltype(P.profile))
+    @cuda threads=Nx blocks=(Ny, Nz) shmem=sz gpu_accumulate_xy!(P.profile, model.poisson_solver.storage, P.field.data, nothing, +)
+    P.profile /= (Nx*Ny)  # Normalize to get the mean from the sum.
+end
+
+####
+#### Horizontally averaged vertical profiles of products (e.g. covariances)
+####
+
+mutable struct ProductProfile{P, F1, F2, I, T} <: Diagnostic
+      profile :: P
+       field1 :: F1
+       field2 :: F2
+    frequency :: I
+     interval :: T
+     previous :: Float64
+end
+
+function ProductProfile(model, field1, field2; frequency=nothing, interval=nothing)
+    freq, int = validate_interval(frequency, interval)
+    profile = zeros(model.arch, model.grid, 1, 1, model.grid.Nz)
+    ProductProfile(profile, field1, field2, freq, int, 0.0)
+end
+
+function run_diagnostic(model::Model, P::ProductProfile{<:Array})
+    # This will allocate lots of memory but we mostly care about GPU performance for now...
+    P.profile .= mean(data(P.field1) .* data(P.field2), dims=[1, 2])
+end
+
+@hascuda function run_diagnostic(model::Model, P::ProductProfile{<:CuArray})
+    Nx, Ny, Nz = P.field1.grid.Nx, P.field1.grid.Ny, P.field1.grid.Nz
+    sz = 2Nx * sizeof(eltype(P.profile))
+    @cuda threads=Nx blocks=(Ny, Nz) shmem=sz gpu_accumulate_xy!(P.profile, model.poisson_solver.storage, P.field1.data, P.field2.data, +)
+    P.profile ./= (Nx*Ny)  # Normalize to get the mean from the sum.
+    return
+end
+
+####
+#### Velocity covariances
+####
+
+mutable struct VelocityCovarianceProfiles{UU, UV, UW, VV, VW, WW, I, T} <: Diagnostic

```

Comment:
It doesn't seem necessary to have a dedicated struct for what is essentially a tuple of diagnostics. Was this created because it is too onerous to simply define 6 individual diagnostics for the velocity covariances? Could we just provide a constructor that creates a tuple of the velocity covariance diagnostics, rather than dedicating a struct to it? It seems like the functionality would be the same.

Reply:
This doesn't have to be a struct I agree. Could just define the product profiles we want (and this struct forces us to compute all six).

Reply:
What do you think about modifying `VelocityCovarianceProfiles` to be a function that returns an array or tuple of 6 velocity covariance diagnostics?